File: code/Python/9-Palindrome Number.py
# Determine whether an integer is a palindrome. Do this without extra space.

# click to show spoilers.

# Some hints:
# Could negative integers be palindromes? (ie, -1)

# If you are thinking of converting the integer to string, note the restriction of using extra space.

# You could also try reversing an integer. However, if you have solved the problem "Reverse Integer", you know that the reversed integer might overflow. How would you handle such case?

# There is a more generic way of solving this problem.

# 因为不能用额外空间，所以不能用专程String的办法，单纯在x上做取余和除法计算，得到首位和末位

class Solution(object):
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        if x < 0:
            return False
        l = 1
        while x/l >= 10:
            l *= 10
        while x:
            if x%10 != x/l:
                return False
            else:
                x = (x % l) / 10
                l /= 100
        return True

        

File: code/C++/9-Palindrome Number .cpp

//Palindrome Number 
//Determine whether an integer is a palindrome. Do this without extra space.
//转换成string之后好做一点


class Solution {
public:
    bool isPalindrome(int x) {
        if(x<0)
          return false;
        string tmp = toString(x);
        return isPalindrome(tmp);
    }
    
    string toString(int x){
        string res;
        while(x>0){
            res.insert(res.begin(), '0'+x%10);
            x=x/10;
        } 
        return res;
    }
    
    bool isPalindrome(string x){
        int i=0, j=x.size()-1;
        while(i<=j){
            if(x[i]!=x[j])
              return false;
            else{
                i++;
                j--;
            }
        }
        return true;
    }
};